### PR TITLE
Aoc day three

### PR DESCRIPTION
**Pull request by DeniseRijborz**

### CODE REVIEW

File: aoc/src/main/scala/aoc2015/Day03.scala [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+    Source
+      .fromFile("aoc/src/resources/aoc2015/Day03")
+      .getLines
+      .mkString
+
+  def generateDirections(): Int =
+    var evenDirections: String = ""
+    var unevenDirections: String = ""
+    lines.zipWithIndex.foreach {
+      case (x, i) =>
```
Comment (nmcb) [inline aoc/src/main/scala/aoc2015/Day03.scala]: syntax preference on: if statements as part of the match cases; and the `+` operator for string concatenation, ie.

```
lines.zipWithIndex.foreach {
  case (x,i) if (i % 2 == 0) => evenDirections = evenDirections + x.toString
  case (x,i)                 => unevenDirections = unevenDirections + x.toString
```

File: aoc/src/main/scala/aoc2015/Day03.scala [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+
+object Day03 extends App:
+  val lines: String =
+    Source
+      .fromFile("aoc/src/resources/aoc2015/Day03")
+      .getLines
+      .mkString
+
+  def generateDirections(): Int =
+    var evenDirections: String = ""
```
Comment (nmcb) [inline aoc/src/main/scala/aoc2015/Day03.scala]: usage of `var`s is frowned upon - see whether you can rewrite the `foreach` loop with a `foldLeft` that calculates the direction strings as immutable values while traversing the lines with indexes

File: aoc/src/main/scala/aoc2015/Day03.scala [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+          unevenDirections = unevenDirections.concat(x.toString)
+    }
+    val even: List[(Int,Int)] = location(0,0,evenDirections)
+    val oneven: List[(Int,Int)] = location(0,0,unevenDirections)
+    val coo: List[(Int,Int)] = even ++ oneven
+    coo.distinct.size
+
+
+  def location(currentx: Int, currenty: Int, directions: String, xAxis: List[Int] = List(0), yAxis: List[Int] = List(0)): List[(Int,Int)] =
+    if (directions.isEmpty) {
```
Comment (nmcb) [inline aoc/src/main/scala/aoc2015/Day03.scala]: perhaps implement the head direction part of the if/else tree with pattern matching on `Char`s, ie. between single-quotes

```
directions.head match
  case '^' =>
    val newx = currentx + 0
    val newy = currenty + 1
    location(newx, newy, directions.tail, xAxis :+ newx, yAxis :+ newy)
  case 'v' =>
    ...
```

**Review by nmcb — CHANGES_REQUESTED** [review]
minor comments - the implementation is correct though not very idiomatically modeled - the problem is an excellent one to learn how modelling helps to make implementational intent of code clearer - you could rewrite the solution eg. employing the following case classes

```
case class Loc(x: Int, y: Int)

case class Area(cur: Loc = Loc(0, 0), houses: Map[Loc,Int] = Map(Loc(0, 0) -> 1))
```

**Comment by DeniseRijborz** [comment]
> minor comments - the implementation is correct though not very idiomatically modeled - the problem is an excellent one to learn how modelling helps to make implementational intent of code clearer - you could rewrite the solution eg. employing the following case classes
> 
> ```
> case class Loc(x: Int, y: Int)
> 
> case class Area(cur: Loc = Loc(0, 0), houses: Map[Loc,Int] = Map(Loc(0, 0) -> 1))
> ```

I changed some code based on the feedback, but I still have to look at using case classes.

**Review by nmcb — CHANGES_REQUESTED** [review]
lgtm - very neat - a little heavy on the updates though; this would be a good one to solve with case classes, eg.the following class models `Pos` and `Area`; and methods `moveTo` and `next` would move the command interpretation into the `Area` class:

```
  case class Loc(x: Int, y: Int)

  case class Area(cur: Loc = Loc(0, 0), houses: Map[Loc,Int] = Map(Loc(0, 0) -> 1)):

    private def moveTo(l: Loc): Area =
      houses.get(l) match
        case Some(c) => copy(cur = l, houses = houses.updated(l, c + 1))
        case None    => copy(cur = l, houses = houses + (l -> 1))

    def next(c: Char): Area =
      c match
        case '>' => { val loc = cur.copy(x = cur.x + 1) ; moveTo(loc) }
        case '<' => { val loc = cur.copy(x = cur.x - 1) ; moveTo(loc) }
        case '^' => { val loc = cur.copy(y = cur.y + 1) ; moveTo(loc) }
        case 'v' => { val loc = cur.copy(y = cur.y - 1) ; moveTo(loc) }
```

an `Area` wraps the current location of Santa (or RoboSanta) and a map of houses containing the current state of presents delivered. the `Char` command list can then be folded over with an initial area, subsequently updating the whole area.

**Comment by nmcb** [comment]
this ☝️ is only one possible implementation ofc - but it shows the use of case classes to model the current state of the problem as a whole.

**Review by nmcb — APPROVED** [review]

**Pull request merged**